src/analysis/calculus/mean_value.lean

```diff
@@ -787,3 +827,84 @@ theorem convex_on_univ_of_deriv2_nonneg {f : ℝ → ℝ} (hf' : differentiable
   convex_on univ f :=
 convex_on_of_deriv2_nonneg convex_univ hf'.continuous.continuous_on hf'.differentiable_on
   hf''.differentiable_on (λ x _, hf''_nonneg x)
+
+/-! ### Functions `f : E → ℝ` -/
+
+/-- Lagrange's Mean Value Theorem, applied to convex domains. -/
+theorem domain_mvt
+  {f : E → ℝ} {s : set E} {x y : E} {f' : E → (E →L[ℝ] ℝ)}
+  (hf : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hs : convex s) (xs : x ∈ s) (ys : y ∈ s) :
+  ∃ z ∈ segment x y, f y - f x = f' z (y - x) :=
+begin
+  have hIccIoo := @Ioo_subset_Icc_self ℝ _ 0 1,
+-- parametrize segment
+  set g : ℝ → E := λ t, x + t • (y - x),
+  have hseg : ∀ t ∈ Icc (0:ℝ) 1, g t ∈ segment x y,
+  { rw segment_eq_image',
+    simp only [mem_image, and_imp, add_right_inj],
+    intros t ht, exact ⟨t, ht, rfl⟩ },
+  have hseg' : Icc 0 1 ⊆ g ⁻¹' s,
+  { rw ← image_subset_iff, unfold image, change ∀ _, _,
+    intros z Hz, rw mem_set_of_eq at Hz, rcases Hz with ⟨t, Ht, hgt⟩,
+    rw ← hgt, exact hs.segment_subset xs ys (hseg t Ht) },
+-- derivative of pullback of f under parametrization
+  have hfg: ∀ t ∈ Icc (0:ℝ) 1, has_deriv_within_at (f ∘ g)
+    ((f' (g t) : E → ℝ) (y-x)) (Icc (0:ℝ) 1) t,
+  { intros t Ht,
+    have hg : has_deriv_at g (y-x) t,
+    { have := ((has_deriv_at_id t).smul_const (y - x)).const_add x,
+      rwa one_smul at this },
+    exact (hf (g t) $ hseg' Ht).comp_has_deriv_within_at _ hg.has_deriv_within_at hseg' },
+-- apply 1-variable mean value theorem to pullback
+  have hMVT : ∃ (t ∈ Ioo (0:ℝ) 1), ((f' (g t) : E → ℝ) (y-x)) = (f (g 1) - f (g 0)) / (1 - 0),
+  { refine exists_has_deriv_at_eq_slope (f ∘ g) _ (by norm_num) _ _,
+    { unfold continuous_on,
+      exact λ t Ht, (hfg t Ht).continuous_within_at },
+    { refine λ t Ht, (hfg t $ hIccIoo Ht).has_deriv_at _,
+      refine mem_nhds_sets_iff.mpr _,
+      use (Ioo (0:ℝ) 1),
+      refine ⟨hIccIoo, _, Ht⟩,
+      simp [real.Ioo_eq_ball, is_open_ball] } },
+-- reinterpret on domain
+  rcases hMVT with ⟨t, Ht, hMVT'⟩,
+  use g t, refine ⟨hseg t $ hIccIoo Ht, _⟩,
+  simp [g, hMVT'],
+end
+
+/-! ### Vector-valued functions `f : E → F`.  Strict differentiability. -/
+
+/-- Over the reals, a continuously differentiable function is strictly differentiable. -/
+lemma strict_fderiv_of_cont_diff
+  {f : E → F} {s : set E}  {x : E} {f' : E → (E →L[ℝ] F)}
+  (hf : ∀ x ∈ s, has_fderiv_within_at f (f' x) s x) (hcont : continuous_on f' s) (hs : s ∈ 𝓝 x) :
+  has_strict_fderiv_at f (f' x) x :=
+begin
+-- turn little-o definition of strict_fderiv into an epsilon-delta statement
+  change is_o _ _ _,
+  apply is_o_iff_forall_is_O_with.mpr,
+  intros c hc,
+  refine is_O_with.of_bound (eventually_iff.mpr (mem_nhds_iff.mpr _)),
+-- the correct ε is the modulus of continuity of f', shrunk to be inside s
+  rcases (metric.continuous_on_iff.mp hcont x (mem_of_nhds hs) c hc) with ⟨ε₁, H₁, hcont'⟩,
+  rcases (mem_nhds_iff.mp hs) with ⟨ε₂, H₂, hε₂⟩,
+  use min ε₁ ε₂, refine ⟨lt_min H₁ H₂, _⟩,
```

Comment:
Sometimes, you write two tactics on a line. This is ok for one liners (or for one liners in big proofs), but not for tactics acting on the main goal of a big proof. So, you should put these on separate lines (or, in this specific case, merge them using 
```suggestion
  refine ⟨min ε₁ ε₂, lt_min H₁ H₂, _⟩,
```
)